experimentation/export_model_predictions.py

```diff
@@ -0,0 +1,154 @@
+import pandas as pd
+from catboost import CatBoostClassifier, Pool
+from nxontology import NXOntology
+from sklearn.pipeline import Pipeline, make_pipeline
+
+from experimentation.model_utils import BiasedMaeMetric
+from nxontology_ml.data import get_efo_otar_slim, read_training_data
+from nxontology_ml.efo import NodeXrefFeatures
+from nxontology_ml.features import NodeInfoFeatures, PrepareNodeFeatures
+from nxontology_ml.model import CatBoostDataFormatter
+from nxontology_ml.text_embeddings.text_embeddings_transformer import (
+    TextEmbeddingsTransformer,
+)
+from nxontology_ml.utils import ROOT_DIR
+
+NON_DISEASE_THERAPEUTIC_AREAS: set[str] = {
+    "EFO:0005932",  # animal_disease
+    "EFO:0001444",  # measurement
+    "EFO:0000651",  # phenotype
+    "GO:0008150",  # biological_process
+    "EFO:0002571",  # medical_procedure
+}
+
+
+def get_disease_nodes(
+    take: int | None = None,
+    nxo: NXOntology[str] | None = None,
+) -> Iterable[str]:
+    """
+    If all roots are in the set of non-disease therapeutic areas (aka roots in EFO OTAR Slim),
+        then its non-disease
+    """
+    nxo = nxo or get_efo_otar_slim()
+    for n in islice(sorted(nxo.graph), take):
+        if any(
+            root not in NON_DISEASE_THERAPEUTIC_AREAS for root in nxo.node_info(n).roots
+        ):
+            yield n
+
+
+rs_cls_to_precision = {
+    "01-disease-subtype": "high",
+    "02-disease-root": "medium",
+    "03-disease-area": "low",
+}
+
+
+@dataclass
+class NodeLabelOutput:
+    """
+    Main output of the model inference tasks
+    """
+
+    identifier: str
+    precision: str
+    proba_high: float
+    proba_medium: float
+    proba_low: float
+    rs_classification: str | None
+    efo_label: str | None
+
+
+def export_model_predictions(
+    export_file: Path = ROOT_DIR / "data/efo_otar_slim_v3.57.0_precisions.tsv",
+    take: int | None = None,
+    text_embeddings_enabled: bool = True,
+) -> None:
+    """
+    1. Train a model (the best performing one) on the entire training set
+    2. Run inference on the entire EFO graph (except the non-disease nodes)
+    3. Export both labels and the feature values for each node
+    """
+    assert not export_file.exists(), f"{export_file} already exists, aborting"
+
+    # 1. Train model
+    X, y = read_training_data(filter_out_non_disease=True, take=take)
+
+    feature_pipeline: Pipeline = make_pipeline(
+        PrepareNodeFeatures(),
+        NodeInfoFeatures(),
+        NodeXrefFeatures(),
+        TextEmbeddingsTransformer.from_config(
+            enabled=text_embeddings_enabled,
+            pca_components=64,
+            use_lda=False,
+        ),
+        CatBoostDataFormatter(),
+    )
+    X_transform = feature_pipeline.fit_transform(X, y)
+    model = CatBoostClassifier(
+        eval_metric=BiasedMaeMetric(),
+        custom_metric=["MultiClass", "AUC", "F1"],
+        learning_rate=0.5,
+        iterations=5000,
+        metric_period=250,
+    )
+    model.fit(
```

Comment:
What was the rationale for refitting the model every time?  Is this deterministic/reproducible?

Reply:
> What was the rationale for refitting the model every time?

That's a great point. 

There isn't a proper rationale: I quickly stopped saving the model weights each time I ran an experiment because it was taking a lot of disk space. For each experiment, I only export (locally) the experiment metadata so it can be visualized in the notebooks. 
Note: There is an underlying assumption that models can be (approximatively) re-created by running the same feature logic on the input records (i.e. the model weights + output labels are an "almost pure" function of the inputs and logic that is saved in git).

Regarding this specific PR: After a discussion with @dhimmel I was under the impression that were mostly interested in snapshotting the feature + output data at this point (as opposed to the model), so the model still isn't saved anywhere.

>  Is this deterministic/reproducible?

Not at the moment. The order of the shuffled records in the training set is consistent but there is randomness in some of the features and CatBoost. It might be quite simple to set a seed everywhere there is randomness, haven't looked into it. It does seem desirable to make this more deterministic and probably to save the model itself.

Reply:
Eventually determinism would be ideal, but for the first release, we have what we need! If the primary model can be saved that could also be nice, but can be done later.